### PR TITLE
Allow to check if sync::Once is already initialized

### PR DESCRIPTION
**Pull request by matklad**
Hi! 

I propose to expose a way to check if a `Once` instance is initialized. 

I need it in `once_cell`. `OnceCell` is effetively a pair of `(Once, UnsafeCell<Option<T>>)`, which can set the `T` only once. Because I can't check if `Once` is initialized, I am forced to add an indirection and check the value of ptr instead:

https://github.com/matklad/once_cell/blob/8127a81976c3f2f4c0860562c3f14647ebc025c0/src/lib.rs#L423-L429

https://github.com/matklad/once_cell/blob/8127a81976c3f2f4c0860562c3f14647ebc025c0/src/lib.rs#L457-L461

The `parking_lot`'s version of `Once` exposes the state as an enum: https://docs.rs/parking_lot/0.6.3/parking_lot/struct.Once.html#method.state.

I suggest, for now, just to add a simple `bool` function: this fits my use-case perfectly, exposes less implementation details, and is forward-compatible with more fine-grained state checking.

### CODE REVIEW

File: src/libstd/sync/once.rs [hunk 1]
```diff
@@ -295,6 +295,47 @@ impl Once {
...
+    /// static INIT: Once = Once::new();
+    ///
+    /// assert_eq!(INIT.is_completed(), false);
+    /// let handle = thread::spawn(|| {
+    ///     INIT.call_once(|| panic!());
+    /// });
+    /// assert!(handle.join().is_err());
+    /// assert_eq!(INIT.is_completed(), false);
+    /// ```
+    #[unstable(feature = "once_is_completed", issue = "42")]
```
Comment (mzji) [inline src/libstd/sync/once.rs]: Isn't this issue number wrong? Issue #42 seems totally irrelevant to this code.

Comment (matklad) [reply src/libstd/sync/once.rs]: It's a placeholder, there's no tracking issue for this yet

**Comment by matklad** [comment]
Here's a somewhat better comparison which shows explains why `.is_completed` would be neat.


| [parking lot](https://github.com/matklad/once_cell/blob/c9b5367e37cdbf45686dee3f1258d28cce30ee95/src/imp_pl.rs#L30-L33) | [std](https://github.com/matklad/once_cell/blob/c9b5367e37cdbf45686dee3f1258d28cce30ee95/src/imp_std.rs#L33-L36)|
|---|--------|

File: src/libstd/sync/once.rs [hunk 2]
```diff
@@ -295,6 +295,47 @@ impl Once {
...
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(once_is_completed)]
+    /// use std::sync::Once;
+    ///
+    /// static INIT: Once = Once::new();
+    ///
+    /// assert_eq!(INIT.is_completed(), false);
```
Comment (shepmaster) [inline src/libstd/sync/once.rs]: Think these would normally be written as `assert!(INIT.is_completed())` /  `assert!(!INIT.is_completed())`

File: src/libstd/sync/once.rs [hunk 3]
```diff
@@ -295,6 +295,47 @@ impl Once {
...
+    /// let handle = thread::spawn(|| {
+    ///     INIT.call_once(|| panic!());
+    /// });
+    /// assert!(handle.join().is_err());
+    /// assert_eq!(INIT.is_completed(), false);
+    /// ```
+    #[unstable(feature = "once_is_completed", issue = "42")]
+    pub fn is_completed(&self) -> bool {
+        self.state.load(Ordering::Acquire) == COMPLETE
```
Comment (shepmaster) [inline src/libstd/sync/once.rs]: I'm never smart enough to use any ordering besides `SeqCst`, but perhaps you can explain why this ordering is appropriate to whoever is clever enough to understand?

**Review by shepmaster — COMMENTED** [review]
This feels like a highly abusable / misusable feature from the outside looking in. With such a function, I can see people wrapping their usages of `Once` with an extra guard "for the fast path". I don't know if that's a real concern or even if it would be an antipattern, but maybe you could chime in on that a bit?

**Comment by shepmaster** [comment]
Passing on to someone smarter...

r? @Kimundi

File: src/libstd/sync/once.rs (hunk 2 above)
Comment (matklad) [reply src/libstd/sync/once.rs]: I think for boolean-returning methods writing an `assert_eq!` in docs specifically helps a bit with readability. Here's an example from result:

https://doc.rust-lang.org/std/result/enum.Result.html#method.is_ok

File: src/libstd/sync/once.rs (hunk 3 above)
Comment (matklad) [reply src/libstd/sync/once.rs]: Good point! Reshuffled the code a bit to reuse an existing comment :-)

**Comment by matklad** [comment]
>With such a function, I can see people wrapping their usages of Once with an extra guard "for the fast path". I don't know if that's a real concern or even if it would be an antipattern, but maybe you could chime in on that a bit?

Interesting point! It is indeed would be an anti pattern! However, I believe there are cases where you really need to check if once is initialized, without actually initializing it. 

For example, a `Once` may guard initialization of some resource, and some part of code might want to assume that the resources is initialized, without the ability to initialized the resource itself. For example, you might want to have a global logger, initialized in main from command-line arguments. To be able to use the logger in the rest of the program safely, you'll need to get an `Option<Logger>` somehow, and then unwrap it. And to do be able to differentiate between `Some` and `None`, you'll need either `is_completed`, or to maintain your own atomic state variable, which is error-prone.

File: src/libstd/sync/once.rs (hunk 1 above)
Comment (mzji) [reply src/libstd/sync/once.rs]: Well, I just found that `42` is the answer to **everything**, however I still want it to be replaced by the true issue ^v^

Comment (matklad) [reply src/libstd/sync/once.rs]: As soon as @rust-lang/libs decides that we indeed want this feature and creates a tracking issue, I'll update the PR.

Comment (mzji) [reply src/libstd/sync/once.rs]: Thanks for the explanation!

**Comment by Kimundi** [comment]
I've also got feature requests for such functionality in the past in `lazy_static`, so it seems like a useful addition.

@rfcbot fcp merge

**Comment by SimonSapin** [comment]
@rfcbot document the race condition

I think there is a potential race condition:

* `is_completed` does its atomic load, and returns `false`
* Then another thread completes initialization
* Then the first thread acts on the boolean return value, which is now outdated.

This race is probably fine in some cases. I think it is in the use case described by https://github.com/rust-lang/rust/pull/53027#issuecomment-410719132. But still, this should be mentioned in the doc-comment to try and make callers aware of it.

**Comment by matklad** [comment]
>But still, this should be mentioned in the doc-comment to try and make callers aware of it.

Yup, added this case to the doc. Ideally, these all should be formulated in terms of "happens before", but I can't find a short way to express that.

As a non-native speaker, I am also not sure whether it should be `is_completed` or `is_complete`.

**Comment by sfackler** [comment]
`has_completed`?

**Comment by RalfJung** [comment]
> Yup, added this case to the doc. Ideally, these all should be formulated in terms of "happens before", but I can't find a short way to express that.

In this case, the intuition I usually try to convey is that the return value is outdated. So people should think of `is_completed` as saying "at some point in the past (when I was checking), the `Once` was [not] initialized". That should hopefully make it clear that it might have been initialized since then.

Maybe it should be called `was_completed`? ;)

**Comment by SimonSapin** [comment]
Oops I messed up the rcfbot command to formally register my concern. But that’s ok, since my concern is resolved now. Thanks!

**Comment by pietroalbini** [comment]
Ping from triage @Kimundi! The FCP ended.

**Comment by TimNN** [comment]
Ping from triage @Kimundi / @rust-lang/libs: This PR requires your review.

**Comment by alexcrichton** [comment]
@bors: r+

I'm gonna go ahead and r+ this to merge but we can of course continue to bikeshed the name while it's unstable!

File: src/libstd/sync/once.rs [hunk 4]
```diff
@@ -295,6 +287,55 @@ impl Once {
...
+    ///
+    /// assert_eq!(INIT.is_completed(), false);
+    /// let handle = thread::spawn(|| {
+    ///     INIT.call_once(|| panic!());
+    /// });
+    /// assert!(handle.join().is_err());
+    /// assert_eq!(INIT.is_completed(), false);
+    /// ```
+    #[unstable(feature = "once_is_completed", issue = "42")]
+    pub fn is_completed(&self) -> bool {
```
Comment (eddyb) [inline src/libstd/sync/once.rs]: `Once` is not generic, so this is missing an `#[inline]`. (@anp found a rayon bench regression likely caused by this)

Comment (anp) [reply src/libstd/sync/once.rs]: https://blog.anp.lol/lolbench-data/benchmarks/rayon-1-0-0-find-serial-find-first.html fwiw

Comment (anp) [reply src/libstd/sync/once.rs]: @eddyb is that documented in the api guidelines or somewhere similar?

Comment (matklad) [reply src/libstd/sync/once.rs]: Fixed in https://github.com/rust-lang/rust/pull/54662
